fili-core/src/main/java/com/yahoo/bard/webservice/data/filterbuilders/DruidInFilterBuilder.java

```diff
@@ -0,0 +1,99 @@
+
+import com.yahoo.bard.webservice.data.dimension.Dimension;
+import com.yahoo.bard.webservice.data.dimension.DimensionRowNotFoundException;
+import com.yahoo.bard.webservice.druid.model.filter.AndFilter;
+import com.yahoo.bard.webservice.druid.model.filter.Filter;
+import com.yahoo.bard.webservice.druid.model.filter.InFilter;
+import com.yahoo.bard.webservice.druid.model.filter.NotFilter;
+import com.yahoo.bard.webservice.web.ApiFilter;
+
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * A {@code DruidInFilterBuilder} builds a conjunction of a in-filter and another in-filter wrapped in a not-filter.
+ * <p>
+ * {@code DruidInFilterBuilder} sees API filters as in two categories
+ * <ol>
+ *     <li>
+ *         positive filters whose API filter operations are one of the following:
+ *         <ul>
+ *             <li> {@link com.yahoo.bard.webservice.web.FilterOperation#in}
+ *             <li> {@link com.yahoo.bard.webservice.web.FilterOperation#startswith}
+ *             <li> {@link com.yahoo.bard.webservice.web.FilterOperation#contains}
+ *             <li> {@link com.yahoo.bard.webservice.web.FilterOperation#eq}
+ *         </ul>
+ *     <li> negative filters whose API filter operation is {@link com.yahoo.bard.webservice.web.FilterOperation#notin}
+ * </ol>
+ * Dimension row values matching the positive API filters are grouped in the single in-filter. Those values matching the
+ * negative API filters are grouped in the other in-filter wrapped in the not-filter.
+ * <p>
+ * For example, suppose a dimension {@code D} has dimension rows {@code 1}, {@code 2}, {@code 3}, {@code 4}, {@code 5}.
+ * A filter clause in in a request like {@code D|id-in[1, 2],D|id-notin[4, 5]} results in an and-filter by
+ * {@code DruidInFilterBuilder}: {@code AND(IN(1, 2), NOT(IN(4, 5)))}. This and-filter is a conjunction of a in-filter
+ * and another in-filter wrapped in a not-filter.
+ * <p>
+ * Compared with {@link DruidOrFilterBuilder}, the advantage of {@code DruidInFilterBuilder} is reducing the number of
+ * Druid filters in a single Druid query. {@code DruidInFilterBuilder} is an enhancement of the
+ * {@link DruidOrFilterBuilder} and is the default Druid filter builder in Fili.
+ */
+public class DruidInFilterBuilder extends ConjunctionDruidFilterBuilder {
+    private static final Logger LOG = LoggerFactory.getLogger(DruidInFilterBuilder.class);
+
+    @Override
+    protected Filter buildDimensionFilter(Dimension dimension, Set<ApiFilter> filters)
+            throws DimensionRowNotFoundException {
+        LOG.trace("Building dimension filter using dimension: {} \n\n and set of filter: {}", dimension, filters);
+
+        List<Filter> inFilters = new ArrayList<>(); // A set with at most two in-filters(positive & negative)
```

Comment:
This isn't used until _much_ later in the class. It's usually a good idea to define variables just before they are used so that they are easier to trace through a method.